packer-windows/conf/bin/bk-install-elastic-stack.ps1

```diff
@@ -0,0 +1,137 @@
+## Installs the Buildkite Agent, run from the CloudFormation template
+
+Set-PSDebug -Trace 2
+
+# Stop script execution when a non-terminating error occurs
+$ErrorActionPreference = "Stop"
+
+function on_error {
+  $errorLine=$_.InvocationInfo.ScriptLineNumber
+  $errorMessage=$_.Exception
+
+  aws autoscaling set-instance-health `
+    --instance-id "(Invoke-WebRequest -UseBasicParsing http://169.254.169.254/latest/meta-data/instance-id).content" `
+    --health-status Unhealthy
+
+  cfn-signal `
+    --region "$Env:AWS_REGION" `
+    --stack "$Env:BUILDKITE_STACK_NAME" `
+    --reason "Error on line ${errorLine}: $errorMessage" `
+    --resource "AgentAutoScaleGroup" `
+    --exit-code 1
+}
+
+trap {on_error}
+
+$Env:INSTANCE_ID=(Invoke-WebRequest -UseBasicParsing http://169.254.169.254/latest/meta-data/instance-id).content
+$DOCKER_VERSION=(docker --version).split(" ")[2].Replace(",","")
+
+$PLUGINS_ENABLED=@()
+If ($Env:SECRETS_PLUGIN_ENABLED -eq "true") { $PLUGINS_ENABLED += "secrets" }
+If ($Env:ECR_PLUGIN_ENABLED -eq "true") { $PLUGINS_ENABLED += "ecr" }
+If ($Env:DOCKER_LOGIN_PLUGIN_ENABLED -eq "true") { $PLUGINS_ENABLED += "docker-login" }
+
+# cfn-env is sourced by the environment hook in builds
+Set-Content -Path C:\buildkite-agent\cfn-env -Value @"
+export DOCKER_VERSION=$DOCKER_VERSION
+export BUILDKITE_STACK_NAME=$Env:BUILDKITE_STACK_NAME
+export BUILDKITE_STACK_VERSION=$Env:BUILDKITE_STACK_VERSION
+export BUILDKITE_AGENTS_PER_INSTANCE=$Env:BUILDKITE_AGENTS_PER_INSTANCE
+export BUILDKITE_SECRETS_BUCKET=$Env:BUILDKITE_SECRETS_BUCKET
+export AWS_DEFAULT_REGION=$Env:AWS_REGION
+export AWS_REGION=$Env:AWS_REGION
+export PLUGINS_ENABLED="$PLUGINS_ENABLED"
+export BUILDKITE_ECR_POLICY=$Env:BUILDKITE_ECR_POLICY
+"@
+
+If ($Env:BUILDKITE_AGENT_RELEASE -eq "edge") {
+  Write-Output "Downloading buildkite-agent edge..."
+  Invoke-WebRequest -OutFile C:\buildkite-agent\bin\buildkite-agent-edge.exe -Uri "https://download.buildkite.com/agent/experimental/latest/buildkite-agent-windows-amd64.exe"
+  buildkite-agent-edge.exe --version
+}
+
+Copy-Item -Path C:\buildkite-agent\bin\buildkite-agent-${Env:BUILDKITE_AGENT_RELEASE}.exe -Destination C:\buildkite-agent\bin\buildkite-agent.exe
+
+$agent_metadata=@(
+  "queue=${Env:BUILDKITE_QUEUE}"
+  "docker=${DOCKER_VERSION}"
+  "stack=${Env:BUILDKITE_STACK_NAME}"
+  "buildkite-aws-stack=${Env:BUILDKITE_STACK_VERSION}"
+)
+
+If (Test-Path Env:BUILDKITE_AGENT_TAGS) {
+  $agent_metadata += $Env:BUILDKITE_AGENT_TAGS.split(",")
+}
+
+$OFS=","
+Set-Content -Path C:\buildkite-agent\buildkite-agent.cfg -Value @"
+name="${Env:BUILDKITE_STACK_NAME}-${Env:INSTANCE_ID}-%n"
+token="${Env:BUILDKITE_AGENT_TOKEN}"
+tags=$agent_metadata
+tags-from-ec2=true
+timestamp-lines=${Env:BUILDKITE_AGENT_TIMESTAMP_LINES}
+hooks-path="C:\buildkite-agent\hooks"
+build-path="C:\buildkite-agent\builds"
+plugins-path="C:\buildkite-agent\plugins"
+experiment="${Env:BUILDKITE_AGENT_EXPERIMENTS}"
+priority=%n
+shell=powershell
```

Comment:
I'd like to know how people feel about using `shell=powershell`. Does anyone think using the default `cmd.exe` would be better?

Reply:
In Windows, Powershell is the choice I'd expect.

Reply:
I guess it depends on what versions of Windows this template supports. If we want to be able to support older versions of windows where Powershell's availability and/or stability is suspect, we might want to make it configurable?

Reply:
I'd prefer `CMD.exe`, just because we don't yet support Powershell officially in the agent.